certdepot/depot_test.go

```diff
@@ -0,0 +1,473 @@
+						assert.Equal(t, name, u.ID)
+						assert.Equal(t, string(certData), u.Cert)
+						assert.Equal(t, string(keyData), u.PrivateKey)
+						assert.Equal(t, string(certReqData), u.CertReq)
+						assert.Equal(t, string(certRevocListData), u.CertRevocList)
+						assert.Equal(t, newTTL, u.TTL)
+					},
+				},
+				{
+					name: "CheckOnExistingUserWithNoData",
+					test: func(t *testing.T, d depot.Depot) {
+						name := "alice"
+						u := &User{
+							ID: name,
+						}
+						require.NoError(t, session.DB(databaseName).C(collectionName).Insert(u))
+
+						assert.False(t, d.Check(depot.CrtTag(name)))
+						assert.False(t, d.Check(depot.PrivKeyTag(name)))
+						assert.False(t, d.Check(depot.CsrTag(name)))
+						assert.False(t, d.Check(depot.CrlTag(name)))
+					},
+				},
+				{
+					name: "GetOnExistingUserWithNoData",
+					test: func(t *testing.T, d depot.Depot) {
+						name := "bob"
+						u := &User{
+							ID:  name,
+							TTL: time.Now(),
+						}
+						require.NoError(t, session.DB(databaseName).C(collectionName).Insert(u))
+
+						data, err := d.Get(depot.CrtTag(name))
+						assert.Error(t, err)
+						assert.Nil(t, data)
+						data, err = d.Get(depot.PrivKeyTag(name))
+						assert.Error(t, err)
+						assert.Nil(t, data)
+						data, err = d.Get(depot.CsrTag(name))
+						assert.Error(t, err)
+						assert.Nil(t, data)
+						data, err = d.Get(depot.CrlTag(name))
+						assert.Error(t, err)
+						assert.Nil(t, data)
+					},
+				},
+				{
+					name: "GetOnExpiredTTL",
+					test: func(t *testing.T, d depot.Depot) {
+						name := "bob"
+						u := &User{
+							ID:            name,
+							Cert:          "cert",
+							PrivateKey:    "key",
+							CertReq:       "certReq",
+							CertRevocList: "certRevocList",
+							TTL:           time.Time{},
+						}
+						require.NoError(t, session.DB(databaseName).C(collectionName).Insert(u))
+
+						data, err := d.Get(depot.CrtTag(name))
+						assert.Error(t, err)
+						assert.Nil(t, data)
+						data, err = d.Get(depot.PrivKeyTag(name))
+						assert.NoError(t, err)
+						assert.Equal(t, u.PrivateKey, string(data))
+						data, err = d.Get(depot.CsrTag(name))
+						assert.NoError(t, err)
+						assert.Equal(t, u.CertReq, string(data))
+						data, err = d.Get(depot.CrlTag(name))
+						assert.Error(t, err)
+						assert.Nil(t, data)
+					},
+				},
+
+				{
+					name: "DeleteWhenDNE",
+					test: func(t *testing.T, d depot.Depot) {
+						name := "bob"
+
+						assert.NoError(t, d.Delete(depot.CrtTag(name)))
+						assert.NoError(t, d.Delete(depot.PrivKeyTag(name)))
+						assert.NoError(t, d.Delete(depot.CsrTag(name)))
+						assert.NoError(t, d.Delete(depot.CrlTag(name)))
+					},
+				},
+			},
+		},
+	} {
+		t.Run(impl.name, func(t *testing.T) {
+			for _, test := range impl.tests {
+				t.Run(test.name, func(t *testing.T) {
+					d := impl.setup()
+					defer impl.cleanup()
+
+					test.test(t, d)
+				})
+			}
+			t.Run("Put", func(t *testing.T) {
```

Comment:
The main thing I'm interested in changing here is that it might make sense to put more sub-tests in these generic cases. Just to break things up and make it more clear what the expected aspect is that you're trying to test. When things break this makes it easier to figure out what happened.